File: POH7/Megane/Main.php
<?php
// Try POH
// author: Leonardone @ NEETSDKASU

function gs()  { return trim(fgets(STDIN)); }
function gi()  { return intval(gs()); }
function gss() { return explode(' ', fgets(STDIN)); }
function gis() { return array_map('intval', gss()); }
function nmapf($n, $f) { return array_map($f, range(0, $n - 1)); }
function ngs($n) { return nmapf($n, 'gs'); }
function ngi($n) { return nmapf($n, 'gi'); }
function ngss($n){ return nmapf($n, 'gss'); }
function ngis($n){ return nmapf($n, 'gis'); }
function for2p($hr,$wr,$pr) { foreach ($hr as $i) { foreach ($wr as $j) { call_user_func($pr, $i, $j); } } }

$n = gi();
$q = ngs($n);
$m = gi();
$p = ngs($m);

for2p(range(0,$n-$m),range(0,$n-$m), function ($i,$j) use ($q,$m,$p) {
    $x = array_reduce(range(0,$m-1), function ($c, $e) use ($i,$j,$q,$m,$p) {
        return $c && $p[$e] === substr($q[$i+$e],$j*2,$m*2-1);
    }, TRUE);
    if ($x) {
        echo "$i $j", PHP_EOL;
    }
});

File: POH7/Megane/Main.pl
# Try POH
# author: Leonardone @ NEETSDKASU
# -----------------------------------------------------------------------------
# use List::Util qw(min max);
# -----------------------------------------------------------------------------
sub gs { chomp(my $s = <>); $s }
sub gi() { int(gs) }
sub gss() { split(/ /, gs)  } #  関数名の後の()はプロトタイプで、 ()をつけるとinline展開されるらしい
sub gis() { map(int, gss) }
sub ngt(&$) { my ($f, $n) = @_; my @r; for (1 .. $n) { push @r, &$f } @r } # プロトタイプをつけなくても引数に関数を渡せるが、ブロック渡しはプロトタイプ必要
sub ngs { ngt \&gs, shift }
sub ngi($) { ngt(\&gi, shift) } #なんとなくプロトタイプをつけてみた、$や@などの変数のタイプ記号を並べるらしい
sub ngss { ngt sub { [gss] }, shift }  # ラムダっぽい動的関数渡し
sub ngis { ngt { [gis] } shift } # ブロック渡し
sub isort { sort {$a <=> $b} @_ } # sort関数内で動的スコープの$aと$bが定義されるからブロック渡しで参照できるとのこと
sub disp2d { for my $y (@_) { for my $x (@{$y}) { print "$x," } print "\n" } }
# -----------------------------------------------------------------------------



sub solve {
    my $n = gi;
    my @p = ngs $n;
    my $m = gi;
    my $q = join '', ngs $m;
    for my $i (0 .. $n - $m) {
        for my $j (0 .. $n - $m) {
            my $u = join '', map { substr $_, $j * 2, $m * 2 - 1 } @p[$i .. $i + $m - 1];
            if ($q eq $u) {
                print "$i $j\n"
            }
        }
    }
}


solve
